{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0358fb69",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\u001b[36mℹ\u001b[39m Loading \u001b[34m\u001b[34mCrossTalkeR\u001b[34m\u001b[39m\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::union’ by ‘igraph::union’ when loading ‘CrossTalkeR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘gtools::permute’ by ‘igraph::permute’ when loading ‘CrossTalkeR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::as_data_frame’ by ‘igraph::as_data_frame’ when loading ‘CrossTalkeR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘dplyr::groups’ by ‘igraph::groups’ when loading ‘CrossTalkeR’”\n",
      "Warning message:\n",
      "“replacing previous import ‘clusterProfiler::simplify’ by ‘igraph::simplify’ when loading ‘CrossTalkeR’”\n",
      "\n",
      "\n",
      "Warning message:\n",
      "“replacing previous import ‘igraph::as_data_frame’ by ‘tibble::as_data_frame’ when loading ‘CrossTalkeR’”\n",
      "Loading required package: dplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:CrossTalkeR’:\n",
      "\n",
      "    groups, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "Loading required package: tidyverse\n",
      "\n",
      "── \u001b[1mAttaching packages\u001b[22m ─────────────────────────────────────── tidyverse 1.3.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mggplot2\u001b[39m 3.3.5     \u001b[32m✔\u001b[39m \u001b[34mpurrr  \u001b[39m 0.3.4\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtibble \u001b[39m 3.1.6     \u001b[32m✔\u001b[39m \u001b[34mstringr\u001b[39m 1.4.0\n",
      "\u001b[32m✔\u001b[39m \u001b[34mtidyr  \u001b[39m 1.2.0     \u001b[32m✔\u001b[39m \u001b[34mforcats\u001b[39m 0.5.1\n",
      "\u001b[32m✔\u001b[39m \u001b[34mreadr  \u001b[39m 2.1.2     \n",
      "\n",
      "── \u001b[1mConflicts\u001b[22m ────────────────────────────────────────── tidyverse_conflicts() ──\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32mcompose()\u001b[39m  masks \u001b[34mCrossTalkeR\u001b[39m::compose()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mtidyr\u001b[39m::\u001b[32mcrossing()\u001b[39m masks \u001b[34mCrossTalkeR\u001b[39m::crossing()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mfilter()\u001b[39m   masks \u001b[34mCrossTalkeR\u001b[39m::filter(), \u001b[34mstats\u001b[39m::filter()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mgroups()\u001b[39m   masks \u001b[34mCrossTalkeR\u001b[39m::groups()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mdplyr\u001b[39m::\u001b[32mlag()\u001b[39m      masks \u001b[34mCrossTalkeR\u001b[39m::lag(), \u001b[34mstats\u001b[39m::lag()\n",
      "\u001b[31m✖\u001b[39m \u001b[34mpurrr\u001b[39m::\u001b[32msimplify()\u001b[39m masks \u001b[34mCrossTalkeR\u001b[39m::simplify()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "devtools::load_all(\"~/Documents/CrossTalkeR/\")\n",
    "require(dplyr)\n",
    "require(tidyverse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9b2e945f",
   "metadata": {},
   "outputs": [],
   "source": [
    "paths <-readRDS('./cc_control.RDS')\n",
    "paths1 <-paths$cellphonedb%>%\n",
    "  filter(pvalue<=0.01) %>%\n",
    "  mutate(gene_A=gsub('_','*',ligand.complex)) %>%\n",
    "  mutate(gene_B=gsub('_','*',receptor.complex)) %>%html\n",
    "  mutate(source=gsub(':|-|/','',source)) %>%\n",
    "  mutate(target=gsub(':|-|/','',target)) %>%\n",
    "  mutate(type_gene_A = rep('Ligand',length(.data$ligand))) %>%\n",
    "  mutate(type_gene_B = rep('Receptor',length(.data$receptor))) \n",
    "  \n",
    "paths <-readRDS('./cc_ADKPD.RDS')\n",
    "paths2 <-paths$cellphonedb %>% \n",
    "  filter(pvalue<=0.01) %>%\n",
    "  mutate(gene_A=gsub('_','*',ligand.complex)) %>%\n",
    "  mutate(gene_B=gsub('_','*',receptor.complex)) %>%\n",
    "  mutate(source=gsub(':|-|/','',source)) %>%\n",
    "  mutate(target=gsub(':|-|/','',target)) %>%\n",
    "  mutate(type_gene_A = rep('Ligand',length(.data$ligand))) %>%\n",
    "  mutate(type_gene_B = rep('Receptor',length(.data$receptor))) \n",
    "\n",
    "teste <- list()\n",
    "teste[['control']] <- paths1 \n",
    "teste[['ADKPD']] <- paths2 \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b99d1d81",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Reading Files\n",
      "\n",
      "Create a Differential Table\n",
      "\n",
      "Calculating CCI Ranking\n",
      "\n",
      "ADKPD_x_control\n",
      "\n",
      "\n",
      "\n",
      "Calculating GCI Ranking\n",
      "\n",
      "ADKPD_x_control\n",
      "\n",
      "Annotating the top Cell Genes\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Reading KEGG annotation online:\n",
      "\n",
      "\n",
      "Reading KEGG annotation online:\n",
      "\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:1 mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "Defining templates\n",
      "\n",
      "\u001b[1m\u001b[22mAdding missing grouping variables: `cellpair`\n",
      "\u001b[1m\u001b[22mAdding missing grouping variables: `cellpair`\n",
      "Generating Report\n",
      "\n",
      "Preparing Single Phenotype Report\n",
      "\n",
      "Loading required package: igraph\n",
      "\n",
      "\n",
      "Attaching package: 'igraph'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:purrr':\n",
      "\n",
      "    compose, simplify\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    crossing\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tibble':\n",
      "\n",
      "    as_data_frame\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:dplyr':\n",
      "\n",
      "    as_data_frame, groups, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:CrossTalkeR':\n",
      "\n",
      "    as_data_frame\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    decompose, spectrum\n",
      "\n",
      "\n",
      "The following object is masked from 'package:base':\n",
      "\n",
      "    union\n",
      "\n",
      "\n",
      "Loading required package: patchwork\n",
      "\n",
      "Loading required package: factoextra\n",
      "\n",
      "Welcome! Want to learn more? See two factoextra-related books at https://goo.gl/ve3WBa\n",
      "\n",
      "Loading required package: ComplexHeatmap\n",
      "\n",
      "Loading required package: grid\n",
      "\n",
      "========================================\n",
      "ComplexHeatmap version 2.8.0\n",
      "Bioconductor page: http://bioconductor.org/packages/ComplexHeatmap/\n",
      "Github page: https://github.com/jokergoo/ComplexHeatmap\n",
      "Documentation: http://jokergoo.github.io/ComplexHeatmap-reference\n",
      "\n",
      "If you use it in published research, please cite:\n",
      "Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional \n",
      "  genomic data. Bioinformatics 2016.\n",
      "\n",
      "The new InteractiveComplexHeatmap package can directly export static \n",
      "complex heatmaps into an interactive Shiny app with zero effort. Have a try!\n",
      "\n",
      "This message can be suppressed by:\n",
      "  suppressPackageStartupMessages(library(ComplexHeatmap))\n",
      "========================================\n",
      "\n",
      "\n",
      "Loading required package: DT\n",
      "\n",
      "\n",
      "Attaching package: 'DT'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:CrossTalkeR':\n",
      "\n",
      "    JS\n",
      "\n",
      "\n",
      "Printing CCI control\n",
      "\n",
      "Warning message in eattrs[[name]][index] <- value:\n",
      "\"number of items to replace is not a multiple of replacement length\"\n",
      "Printing CCI ADKPD\n",
      "\n",
      "Warning message in eattrs[[name]][index] <- value:\n",
      "\"number of items to replace is not a multiple of replacement length\"\n",
      "Loading required package: ggrepel\n",
      "\n",
      "Warning message:\n",
      "\"Removed 3337 rows containing missing values (geom_label_repel).\"\n",
      "Warning message:\n",
      "\"ggrepel: 11 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"Removed 3836 rows containing missing values (geom_label_repel).\"\n",
      "Warning message:\n",
      "\"ggrepel: 9 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"ggrepel: 11 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"ggrepel: 9 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"The closing backticks on line 22 (\"     ```\") in Single_Condition_cgi.Rmd do not match the opening backticks \"```\" on line 4. You are recommended to fix either the opening or closing delimiter of the code chunk to use exactly the same numbers of backticks and same level of indentation (or blockquote).\"\n",
      "Warning message:\n",
      "\"The closing backticks on line 44 (\"     ```\") in Single_Condition_cgi.Rmd do not match the opening backticks \"```\" on line 26. You are recommended to fix either the opening or closing delimiter of the code chunk to use exactly the same numbers of backticks and same level of indentation (or blockquote).\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Single_report_files/figure-latex/unnamed-chunk-9-1.png\" but not available.\n",
      "\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Single_report_files/figure-latex/unnamed-chunk-10-1.png\" but not available.\n",
      "\n",
      "Warning message in instance$preRenderHook(instance):\n",
      "\"It seems your data is too big for client-side DataTables. You may consider server-side processing: https://rstudio.github.io/DT/server.html\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Single_report_files/figure-latex/unnamed-chunk-11-1.png\" but not available.\n",
      "\n",
      "Warning message in instance$preRenderHook(instance):\n",
      "\"It seems your data is too big for client-side DataTables. You may consider server-side processing: https://rstudio.github.io/DT/server.html\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Single_report_files/figure-latex/unnamed-chunk-12-1.png\" but not available.\n",
      "\n",
      "Preparing Comparative Phenotype Report\n",
      "\n",
      "Loading required package: ggraph\n",
      "\n",
      "Loading required package: graphlayouts\n",
      "\n",
      "Loading required package: pals\n",
      "\n",
      "Loading required package: extrafont\n",
      "\n",
      "Registering fonts with R\n",
      "\n",
      "Loading required package: colorBlindness\n",
      "\n",
      "Printing CCI ADKPD_x_control\n",
      "\n",
      "Warning message in head_of(all_data@graphs[[curr]], E(all_data@graphs[[curr]]))$name == :\n",
      "\"longer object length is not a multiple of shorter object length\"\n",
      "Warning message in tail_of(all_data@graphs[[curr]], E(all_data@graphs[[curr]]))$name == :\n",
      "\"longer object length is not a multiple of shorter object length\"\n",
      "Warning message in eattrs[[name]][index] <- value:\n",
      "\"number of items to replace is not a multiple of replacement length\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_text_repel).\"\n",
      "Warning message:\n",
      "\"Removed 1 rows containing missing values (geom_segment).\"\n",
      "Warning message:\n",
      "\"ggrepel: 6 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"ggrepel: 6 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"The closing backticks on line 25 (\"     ```\") in Comparative_Condition_cci.Rmd do not match the opening backticks \"```\" on line 6. You are recommended to fix either the opening or closing delimiter of the code chunk to use exactly the same numbers of backticks and same level of indentation (or blockquote).\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Comparative_report_files/figure-latex/unnamed-chunk-8-1.png\" but not available.\n",
      "\n",
      "Warning message:\n",
      "\"Removed 4460 rows containing missing values (geom_label_repel).\"\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"ggrepel: 350 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"ggrepel: 350 unlabeled data points (too many overlaps). Consider increasing max.overlaps\"\n",
      "Warning message:\n",
      "\"The closing backticks on line 23 (\"     ```\") in Comparative_Condition_cgi.Rmd do not match the opening backticks \"```\" on line 4. You are recommended to fix either the opening or closing delimiter of the code chunk to use exactly the same numbers of backticks and same level of indentation (or blockquote).\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Comparative_report_files/figure-latex/unnamed-chunk-15-1.png\" but not available.\n",
      "\n",
      "Warning message in instance$preRenderHook(instance):\n",
      "\"It seems your data is too big for client-side DataTables. You may consider server-side processing: https://rstudio.github.io/DT/server.html\"\n",
      "The magick package is required to crop \"/home/james/sciebo/Sikander2022/Comparative_report_files/figure-latex/unnamed-chunk-16-1.png\" but not available.\n",
      "\n",
      "DONE!\n",
      "\n"
     ]
    }
   ],
   "source": [
    "data <- generate_report(lrpaths = teste,\n",
    "                        genes=NULL,\n",
    "                        out_path='~/sciebo/Sikander2022/',\n",
    "                        threshold=0,\n",
    "                        out_file = 'report.pdf',\n",
    "                        output_fmt = \"pdf_document\",\n",
    "                        report = T,\n",
    "                        sel_columns = c('source','target','gene_A','gene_B','type_gene_A','type_gene_B','lr.mean'))\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
